pkg/logql/log/parser.go

```diff
@@ -348,6 +348,92 @@ func (l *PatternParser) Process(_ int64, line []byte, lbs *LabelsBuilder) ([]byt
 
 func (l *PatternParser) RequiredLabelNames() []string { return []string{} }
 
+type LogfmtExpressionParser struct {
+	/* dec  *logfmt.Decoder
+	keys internedStringSet */
+	expressions map[string][]interface{}
+	keys        internedStringSet
+}
+
+func NewLogfmtExpressionParser(expressions []LogfmtExpression) (*LogfmtExpressionParser, error) {
+	paths := make(map[string][]interface{})
+
+	for _, exp := range expressions {
+		path, err := logfmt.Parse(exp.Expression, false)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse expression [%s]: %w", exp.Expression, err)
+		}
+
+		if !model.LabelName(exp.Identifier).IsValid() {
+			return nil, fmt.Errorf("invalid extracted label name '%s'", exp.Identifier)
+		}
+		paths[exp.Identifier] = path
+	}
+	return &LogfmtExpressionParser{
+		expressions: paths,
+		keys:        internedStringSet{},
+	}, nil
+}
+
+func (l *LogfmtExpressionParser) Process(_ int64, line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+
+	if lbs.ParserLabelHints().NoLabels() {
+		return line, true
+	}
+	// Convert LogfmtExpression to Logfmt type to parse the key value pairs
+	j := NewLogfmtParser()
+	originalLabels := NewBaseLabelsBuilder().ForLabels(lbs.labels(), lbs.labels().Hash())
+	originalLabels.Reset()
+
+	j.dec.Reset(line)
+
+	for j.dec.ScanKeyval() {
+		key, ok := j.keys.Get(j.dec.Key(), func() (string, bool) {
+			sanitized := sanitizeLabelKey(string(j.dec.Key()), true)
+			if !lbs.ParserLabelHints().ShouldExtract(sanitized) {
+				return "", false
+			}
+			if len(sanitized) == 0 {
+				return "", false
+			}
+			if lbs.BaseHas(sanitized) {
+				sanitized = sanitized + duplicateSuffix
+				lbs.Set(sanitized, string(j.dec.Value()))
```

Comment:
This does mean that the `<label>_extracted` label is always added, even though it is not part of the expressions? I would not expect this behaviour. See also `pkg/logql/log/parser_test.go#L838`.

Reply:
It will only be added in case there's a label naming collision, which is only in this use case among all of them:
```
{
  "label override",
  testLine,
  []LogfmtExpression{
	 NewLogfmtExpr("app", "app"),
  },
  labels.Labels{
        {Name: "app", Value: "bar"},
  },
  labels.Labels{
       {Name: "app", Value: "bar"},
       {Name: "app_extracted", Value: "foo"},
  },
},
```